snakemake/remote/GS.py

```diff
@@ -125,6 +127,7 @@ def size(self):
         else:
             return self._iofile.size_local
 
+    @retry.Retry()
```

Comment:
Do you think we need the internal retry here as well?

Reply:
They are different purposes - the internal one is for the case that the download "works" but the hash is wrong. The external one is for connectivity. If the API is working but there is some blip so my hash is wrong, I would want the internal retry to exist, because otherwise I'd just see an error that the hashes don't match (and the user would likely say "Why can't we retry?") If the API is haven't issues, that would raise the exception caught by the external decorator. So I think we are better having both.

Reply:
Yes, but you can simply pass a custom [predicate](https://googleapis.dev/python/google-api-core/latest/retry.html#google.api_core.retry.Retry), can't you? Pass a function that also checks for the exception caused by the hash mismatch (we should add a special exception for that in snakemake/exceptions.py), in addition to what is [default](https://googleapis.dev/python/google-api-core/latest/retry.html#google.api_core.retry.if_transient_error) now.

Reply:
@johanneskoester here are some implementation details:

## Adding CheckSumMismatchException
I first modified the download function to instead of having a loop with retry, to just raise a new exception (defined in snakemake/exceptions.py) for CheckSumMismatchException). Then that would be caught by the decorator. Having the download function raise a special exception type for this case of strings mismatching seems like a logical way to have the predicate know exactly that it happened.

## Writing a Google Storage decorator
I had first started writing the decorator function in snakemake/executors.py, but since we require the retry.is_transient_error function, this would mean trying to do an import in executors.py that might not be relevant for the user (e.g. a Tibanna executor user doesn't care about this Google decorator but it would trigger an error and tell them they need the library). We could move the import into the function, but then we run a similar issue of trying to run the function for some reason without the library, and having the error trigger internal to the check. I thought since we already do this check and then potential raise of import error for the retry library, it would be more straight forward to define the decorator directly in remotes/GS.py, and have it use the CheckSumMismatchException from exceptions.py. The decorator is very simple - it just checks for transient error or this particular kind of new exception. Take a look and let me know if this is what you had in mind!